source/common/tcp/conn_pool.cc

```diff
@@ -400,6 +409,10 @@ ConnPoolImpl::ActiveConn::~ActiveConn() {
   parent_.host_->cluster().resourceManager(parent_.priority_).connections().dec();
 
   parent_.onConnDestroyed(*this);
+  if (!connection_succeeded_) {
```

Comment:
instead of maintaining this extra state I think you can decrement when we get a connect failure, see https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/conn_pool.cc#L145

Reply:
I started out by doing that, but decided to put it in the destructor instead because I wanted to be 100% certain that the counter would always be decremented. I don't really like this boolean either though.

Reply:
@julia-stripe I think we should validate that (whether or not the suggested approach has flaws to be not 100% correct) via some tests?

Reply:
The reason I think this approach makes more sense is that it's more refactor-proof -- we need to decrement the counter in at least 3 cases ([1](https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/conn_pool.cc#L149), [2](https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/conn_pool.cc#L142), [3](https://github.com/envoyproxy/envoy/blob/master/source/common/http/http1/conn_pool.cc#L138)). So if we decremented the counter in every one of those places, it would be easy for someone to accidentally forget to decrement the counter if they added another connection failure case. Putting it in the destructor means that it'll always be decremented no matter what. And it'll still be decremented if an exception is thrown.

Is your concern that:

* the `connection_succeeded_` boolearn isn't atomic? (happy to make it atomic -- this is a pretty slow path so I don't think it should be an issue)
* every connection is storing an extra byte?
* something else?